src/Aspire.Hosting.Azure.Network/AzureVirtualNetworkResource.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Azure.Provisioning.Network;
+using Azure.Provisioning.Primitives;
+
+namespace Aspire.Hosting.Azure;
+
+/// <summary>
+/// Represents an Azure Virtual Network resource.
+/// </summary>
+/// <param name="name">The name of the resource.</param>
+/// <param name="configureInfrastructure">Callback to configure the Azure Virtual Network resource.</param>
+public class AzureVirtualNetworkResource(string name, Action<AzureResourceInfrastructure> configureInfrastructure)
+    : AzureProvisioningResource(name, configureInfrastructure)
+{
+    internal List<AzureSubnetResource> Subnets { get; } = [];
+
+    /// <summary>
+    /// Gets the "id" output reference from the Azure Virtual Network resource.
+    /// </summary>
+    public BicepOutputReference Id => new("id", this);
+
+    /// <summary>
+    /// Gets the "name" output reference for the resource.
+    /// </summary>
+    public BicepOutputReference NameOutput => new("name", this);
+
+    /// <inheritdoc/>
+    public override ProvisionableResource AddAsExistingResource(AzureResourceInfrastructure infra)
```

Comment:
For my education: why return `ProvisionableResource` instead of returning `VirtualNetwork` (which is a `ProvisionableResource`)? Is it because that would not work to  to "override" because "C# does not allow covariant return types for normal methods"?

The reason I am asking is that based on my (rather limited) usage of this API, folks end up upper-casting the result of `AddAsExistingResource`. And this feels so unnecessary - and not type-safe  :(

Reply:
Perhaps not the most elegant but something like... 

```suggestion
    public override ProvisionableResource AddAsExistingResource(AzureResourceInfrastructure infra) =>  AddAsExistingResource(AzureResourceInfrastructure infra, out var _);
    
    public ProvisionableResource AddAsExistingResource(AzureResourceInfrastructure infra, out VirtualNetwork existingVNet)
```

Reply:
>  Is it because that would not work to to "override" because "C# does not allow covariant return types for normal methods"?

Correct. Since this is an override the signature needs to match with the base signature.